// @copilotkit/runtime-client-gql/dist/chunk-UTB4B2W3.mjs
import {
  MessageRole
} from "./chunk-OS54FOSV.mjs";

// src/client/types.ts
import { randomId } from "@copilotkit/shared";
var Message = class {
  constructor(props) {
    props.id ?? (props.id = randomId());
    props.status ?? (props.status = { code: "Success" /* Success */ });
    props.createdAt ?? (props.createdAt = /* @__PURE__ */ new Date());
    Object.assign(this, props);
  }
};
var Role = MessageRole;
var TextMessage = class extends Message {
  constructor(props) {
    super(props);
  }
};
var ActionExecutionMessage = class extends Message {
  constructor(props) {
    super(props);
  }
};
var ResultMessage = class extends Message {
  constructor(props) {
    super(props);
  }
  static decodeResult(result) {
    try {
      return JSON.parse(result);
    } catch (e) {
      return result;
    }
  }
  static encodeResult(result) {
    if (result === void 0) {
      return "";
    } else if (typeof result === "string") {
      return result;
    } else {
      return JSON.stringify(result);
    }
  }
};
var AgentStateMessage = class extends Message {
};

export {
  Message,
  Role,
  TextMessage,
  ActionExecutionMessage,
  ResultMessage,
  AgentStateMessage
};
